### PR TITLE
ESLint take 2

### PR DESCRIPTION
**Pull request by ChristianMurphy**
Second take on #3182

This PR will only enable and enforce recommended rules.
Other style rules are warned but are not yet enforced.
Each of the rules currently warned can be enforced later, one at a time, in separate PRs.

Hopefully this approach will make this process more manageable. :smile: 
and will not torch the tests :frowning:

### CODE REVIEW

**Comment by ChristianMurphy** [comment]
@vkarpov15 okay I'm a little worried now. :worried: 
I haven't changed any javascript yet on this PR and I am getting failing unit tests on [node 0.12](https://travis-ci.org/Automattic/mongoose/jobs/72148617#L294) and [node 0.10](https://travis-ci.org/Automattic/mongoose/jobs/72148618#L284). :confused:

**Comment by vkarpov15** [comment]
@ChristianMurphy don't worry about it too much, the Travis build is quite flakey. I only use it as a general rule of thumb.

**Comment by ChristianMurphy** [comment]
@vkarpov15 question

in this
``` js
/**
 * Sets the readPreference option for the aggregation query.
 *
 * ####Example:
 *
 *     Model.aggregate(..).read('primaryPreferred').exec(callback)
 *
 * @param {String} pref one of the listed preference options or their aliases
 * @param {Array} [tags] optional tags for this query
 * @see mongodb http://docs.mongodb.org/manual/applications/replication/#read-preference
 * @see driver http://mongodb.github.com/node-mongodb-native/driver-articles/anintroductionto1_1and2_2.html#read-preferences
 */

Aggregate.prototype.read = function (pref) {
  if (!this.options) this.options = {};
  read.apply(this, arguments);
  return this;
};
```

is `pref` unused, or is there some magic with `this` and `apply()` happening so that `pref` is used?

Also there is a reference to optional param `tags` in the doc comment, is that something which could/should be removed?

**Comment by vkarpov15** [comment]
Also can you make remove the lint from posttest for now? It makes Travis logs difficult to read and we know there's a lot of lint errors :)

**Comment by ChristianMurphy** [comment]
I'll remove it from "posttest" for now.
I would recommend re-enabling it later, either fully or in quiet mode (which only shows errors not warnings).

Currently quiet mode takes the log from ~13,000 to ~600 lines of output from eslint.

**Comment by ChristianMurphy** [comment]
alright remove eslint from posttest, the logs are looking much nicer :smile:

File: test/common.js [hunk 1]
```diff
@@ -138,7 +136,7 @@ function dropDBs(done) {
         if (!mongos) return done();
 
 
-        var db = start({ uri: mongos, mongos: true });
+        var db = mongoose.connect(mongos, {mongos: true });
```
Comment (ChristianMurphy) [inline test/common.js]: I assume this was trying to connect to the database.
It is not defined elsewhere in the file.

File: test/document.modified.test.js [hunk 2]
```diff
@@ -84,31 +75,27 @@ describe('document modified', function(){
...
     it('reset after save', function(done){
       var db = start()
-        , B = db.model(modelName, collection)
-        , pending = 2;
+        , B = db.model(modelName, collection);
 
       var b = new B;
 
       b.numbers.push(3);
       b.save(function (err) {
```
Comment (ChristianMurphy) [inline test/document.modified.test.js]: Original test code had a race condition, this was contributing to some of the flakiness of Travis CI tests.

**Comment by ChristianMurphy** [comment]
@vkarpov15 
ESLint passes running `eslint . --quiet` with current settings. 
I'm not spotting any new :fire: in the tests,
Travis CI tests on this branch have the same errors as Master,
running the tests on my local machine has flakier than Travis CI :frowning: .

So it would probably be a good idea to run the tests on a different machine,
just to be sure no :bug: were introduced.

But it should be good to go. :smile:

**Comment by vkarpov15** [comment]
@ChristianMurphy sorry, forgot about this. Can you see what the merge conflict is about?

**Comment by ChristianMurphy** [comment]
@vkarpov15 resolved the merge conflicts. :+1:

**Pull request merged**